sd-dhcp-server: remove lease with hashmap_remove_value()

yuwata opened the pull request:
Fixes #22253.

yuwata commented:
@bluca I added one more commit. Could you review again?

yuwata commented:
Note, the issue is caused by b713a99b1a04507412b387214a26e43a2916bd01 (v250). Please backport this to v250-stable.

yuwata commented:
@bluca Thank you!

bluca commented:
Could this be related?

```
17:36:40 ======================================================================
17:36:40 FAIL: test_activation_policy (__main__.NetworkdNetworkTests) (iteration=1, expect_up=True, test='down')
17:36:40 ----------------------------------------------------------------------
17:36:40 Traceback (most recent call last):
17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 3076, in _test_activation_policy
17:36:40     self.wait_operstate('test1', operstate, setup_state=setup_state, setup_timeout=20)
17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 625, in wait_operstate
17:36:40     self.fail(f'Timed out waiting for {link} to reach state {operstate}/{setup_state}')
17:36:40 AssertionError: Timed out waiting for test1 to reach state routable/configured
```

evverx commented:
@yuwata could you add a test? I'm not sure that particular scenario can be easily reproduced with the integration test but I think it should be possible to do that in `test-dhcp-server`.

evverx commented:
Looking at https://coveralls.io/builds/45954594/source?filename=src%2Flibsystemd-network%2Fsd-dhcp-server.c#L903 it seems it isn't covered by any tests. @mrc0mmand just out of curiosity I wonder if `systemd-networkd-tests.py` is included in that coverage report? Given that it's suspiciously green I suspect it is but just to be sure :-)

yuwata commented:
> Could this be related?
> 
> ```
> 17:36:40 ======================================================================
> 17:36:40 FAIL: test_activation_policy (__main__.NetworkdNetworkTests) (iteration=1, expect_up=True, test='down')
> 17:36:40 ----------------------------------------------------------------------
> 17:36:40 Traceback (most recent call last):
> 17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 3076, in _test_activation_policy
> 17:36:40     self.wait_operstate('test1', operstate, setup_state=setup_state, setup_timeout=20)
> 17:36:40   File "/build/./test/test-network/systemd-networkd-tests.py", line 625, in wait_operstate
> 17:36:40     self.fail(f'Timed out waiting for {link} to reach state {operstate}/{setup_state}')
> 17:36:40 AssertionError: Timed out waiting for test1 to reach state routable/configured
> ```

Filed at #22267.

yuwata commented:
Updated.
- Previously separated two commits is merged, as both commits are necessary to fix the issue.
- Added tests for static lease.
- Added several cleanups.

PTAL.

mrc0mmand commented:
> Looking at https://coveralls.io/builds/45954594/source?filename=src%2Flibsystemd-network%2Fsd-dhcp-server.c#L903 it seems it isn't covered by any tests. @mrc0mmand just out of curiosity I wonder if `systemd-networkd-tests.py` is included in that coverage report? Given that it's suspiciously green I suspect it is but just to be sure :-)

It should be:
 * https://github.com/systemd/systemd-centos-ci/blob/750c38923d4046f4bf98fe86f9ccf1f21b700d42/vagrant/test_scripts/test-arch-coverage.sh#L193
 * https://jenkins-systemd.apps.ocp.ci.centos.org/job/upstream-cron-build/lastSuccessfulBuild/artifact/systemd-centos-ci/artifacts_all/artifacts_w3hojuz2/vagrant-logs.ZLQ/vagrant-arch-coverage-testsuite.STO/lcov_networkd_collect_coverage_PASS.log

Hopefully it's doing its job :-)

evverx commented:
@yuwata thanks! I'm not sure the test can trigger the crash with the first two commits reverted. I wonder if the idea was to cover something else?

> Hopefully it's doing its job :-)

@mrc0mmand it's going strong as far as I can see :-)

yuwata commented:
I noticed that the first commit just hide the issue.

yuwata commented:
The previous version just hides the issue, now the root of the issue is also fixed. PTAL!

yuwata commented:
This becomes huge than I expected, but most commits are trivial cleanups.

evverx commented:
@yuwata as far as I can see, the fuzzer kind of covers `lease->server->bound_leases_by_*` but `static_leases_by_*` are always empty there. Could you add them to the fuzzer to let it cover that part as well? It would be even better if they could expire there.

yuwata commented:
@evverx Done. PTAL.

evverx commented:
@yuwata thanks! CIFuzz is kind of broken now so it didn't actually run the fuzzer but I built it manually and ran for about 5 minutes under ASan/UBSan/MSan and it seems to be working. The PR looks good to me but since the PR is different from what was tested in https://github.com/systemd/systemd/issues/22253 I'd probably wait for @pfactum to confirm that the issue is gone (and new issues aren't introduced :-)).

yuwata commented:
Note without the above fixes for sd-dhcp-server, the test https://github.com/systemd/systemd/pull/22254/commits/5e61672762701553ca4370d04b2020c2ec7cc88f with small change below can reproduce the original issue.
```diff
diff --git a/src/libsystemd-network/test-dhcp-server.c b/src/libsystemd-network/test-dhcp-server.c
index c3d18ad825..912a8ddee6 100644
--- a/src/libsystemd-network/test-dhcp-server.c
+++ b/src/libsystemd-network/test-dhcp-server.c
@@ -190,22 +190,22 @@ static void test_message_handler(void) {
         /* request address reserved for static lease (unmatching client ID) */
         test.option_client_id.id[6] = 'H';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 42);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == 0);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
 
         /* request unmatching address */
         test.option_client_id.id[6] = 'G';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 41);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == 0);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
 
         /* request matching address */
         test.option_client_id.id[6] = 'G';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 42);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) < 0);
 
         /* try again */
         test.option_client_id.id[6] = 'G';
         test.option_requested_ip.address = htobe32(INADDR_LOOPBACK + 42);
-        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) == DHCP_ACK);
+        assert_se(dhcp_server_handle_message(server, (DHCPMessage*)&test, sizeof(test)) < 0);
 }
 
 static uint64_t client_id_hash_helper(DHCPClientId *id, uint8_t key[HASH_KEY_SIZE]) {
```

pfactum commented:
Do I apply all 19 commits for my test, or there's some minimal subset that fixes the issue I've reported?

yuwata commented:
> Do I apply all 19 commits for my test, or there's some minimal subset that fixes the issue I've reported?

The commits 8e933ee81b01c95b5c596a4e7229235795c6f0da..43227d5c553a169febf11e323c994bba445350a9 are necessary.
But there exists one conflict when these are applied to v250. I prepared a version which can be applied to v250.
Please try https://github.com/systemd/systemd-stable/pull/160.

yuwata commented:
@pfactum If you also want to backport the reproducer of the issue, then try 5e61672.
With the commits in https://github.com/systemd/systemd-stable/pull/160, the test should pass, but the without the fixes, the test should fail. See https://github.com/systemd/systemd/pull/22254#issuecomment-1023857704 for more detail.

pfactum commented:
With https://github.com/systemd/systemd-stable/pull/160 the crash is not reproducible for me.

yuwata commented:
@pfactum Thank you for your help!

bluca commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1133,14 +1133,17 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message, siz
+                        /* The requested address is in the pool. */
+
+                        if (existing_lease && existing_lease->address != address)
+                                /* We previously assign an address, but the client requests another address. Refusing. */
```
"We previously assigned an address, but the client rested another one. Refuse."

bluca commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1084,9 +1084,13 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message, siz
+                        /* Found a static lease for the client ID. */
+
+                        if (static_lease->address != address)
+                                /* The client requests an address which is different from the static lease. Refusing. */
```
"The client requested an address..."
"Refuse."

bluca commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -612,16 +612,24 @@ static int server_send_offer_or_ack(
         size_t offset;
         int r;
 
+        if (!init_reboot)
```
Feel free to ignore, but to me it seemed better with this condition checked outside of the function - it doesn't seem very related. But not too important.

yuwata replied:
I'd like to keep this as is. This makes `dhcp_server_handle_message()` simpler.

Instead, I added a brief comment why we silently ignore the request in that case.

anitazha reviewed (approved):
LGTM

yuwata commented:
@bluca Thank you. The comments are updated. Upgrading the green label.
